src/librbd/ImageCtx.cc

```diff
@@ -633,29 +635,46 @@ namespace librbd {
   }
 
   int ImageCtx::invalidate_cache() {
-    if (!object_cacher)
-      return 0;
-    flush_async_operations();
+    C_SaferCond ctx;
+    invalidate_cache(&ctx);
+    return ctx.wait();
+  }
+
+  void ImageCtx::invalidate_cache(Context *on_finish) {
+    if (object_cacher == NULL) {
+      on_finish->complete(0);
+      return;
+    }
+
     cache_lock.Lock();
     object_cacher->release_set(object_set);
     cache_lock.Unlock();
-    int r = flush_cache();
+
+    flush_cache_aio(new FunctionContext(boost::bind(
+      &ImageCtx::invalidate_cache_completion, this, _1, on_finish)));
+  }
+
+  void ImageCtx::invalidate_cache_completion(int r, Context *on_finish) {
     if (r == -EBLACKLISTED) {
       Mutex::Locker l(cache_lock);
```

Comment:
looks like this ends up being called with cache_lock already held from the ObjectCacher::flush_set() context